msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenByManagedIdentitySupplier.java

```diff
@@ -63,23 +65,37 @@ AuthenticationResult execute() throws Exception {
                         this.clientApplication,
                         silentRequest);
 
-                return supplier.execute();
+                AuthenticationResult result = supplier.execute();
+                cacheRefreshReason = SilentRequestHelper.NeedsRefresh(
+                        parameters,
+                        result,
+                        LOG);
+
+                // If the token does not need a refresh, return the cached token
+                // Else refresh the token if it is either expired, proactively refreshable, or if the claims are passed.
+                if (cacheRefreshReason == CacheRefreshReason.NOT_APPLICABLE) {
+                    LOG.debug("Returning token from cache");
+                    result.metadata().tokenSource(TokenSource.CACHE);
+                    return result;
+                } else {
+                    LOG.debug(String.format("Refreshing access token. Cache refresh reason: %s", cacheRefreshReason));
```

Comment:
suggestion: When I first saw this I was a bit confused about where the access token would be refreshed, until I realized it'd be taken care of by the return down near line 95.

For better clarity about the code flow and returns. I'd suggest:
- Add an additional `return fetchNewAccessTokenAndSaveToCache()` here in the else
- Rather than have a negative if statement (`if (!managedIdentityParameters.forceRefresh)`) contain all this other logic, cut the execution short with a ` if (managedIdentityParameters.forceRefresh) { return fetchNewAccessToken... }` and have all the other logic come after that 
  - Like it says in the `SilentRequestHelper.needsRefresh()`, if `forceRefresh` is true then there's no need to check any other refresh option

Reply:
I updated it, checking the force refresh in the beginning since force refresh do not need a cache lookup.